File: golem-go-client/bootstrap.go
package golemclient

import (
	"context"
	"crypto/sha256"
	"crypto/subtle"
	"crypto/tls"
	"crypto/x509"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

const certificateHashSHA256 = "sha-256"

// RealtimeConfig is the client-side shape served by golem.Server.RealtimeConfigHandler.
type RealtimeConfig struct {
	Transport               TransportKind
	URL                     string
	ServerCertificateHashes []CertificateHash
	EventualAckIntervalMs   int
}

type realtimeConfigResponse struct {
	Transport               string `json:"transport"`
	URL                     string `json:"url"`
	ServerCertificateHashes []struct {
		Algorithm string `json:"algorithm"`
		Value     string `json:"value"`
	} `json:"serverCertificateHashes"`
	EventualAckIntervalMs *int `json:"eventualAckIntervalMs"`
}

// FetchRealtimeConfig loads and decodes a realtime config JSON endpoint.
func FetchRealtimeConfig(ctx context.Context, endpoint string, client *http.Client) (RealtimeConfig, error) {
	if client == nil {
		client = http.DefaultClient
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return RealtimeConfig{}, fmt.Errorf("golem-go-client: creating realtime config request: %w", err)
	}
	resp, err := client.Do(req)
	if err != nil {
		return RealtimeConfig{}, fmt.Errorf("golem-go-client: fetching realtime config: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
		snippet, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		if len(snippet) > 0 {
			return RealtimeConfig{}, fmt.Errorf("golem-go-client: fetching realtime config: status %d body=%q", resp.StatusCode, string(snippet))
		}
		return RealtimeConfig{}, fmt.Errorf("golem-go-client: fetching realtime config: status %d", resp.StatusCode)
	}
	var body realtimeConfigResponse
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return RealtimeConfig{}, fmt.Errorf("golem-go-client: decoding realtime config: %w", err)
	}
	cfg := RealtimeConfig{
		Transport: TransportKind(body.Transport),
		URL:       body.URL,
	}
	if body.EventualAckIntervalMs != nil {
		cfg.EventualAckIntervalMs = *body.EventualAckIntervalMs
	}
	for _, hash := range body.ServerCertificateHashes {
		decoded, err := decodeCertificateHash(hash.Algorithm, hash.Value)
		if err != nil {
			return RealtimeConfig{}, err
		}
		cfg.ServerCertificateHashes = append(cfg.ServerCertificateHashes, decoded)
	}
	return cfg, nil
}

// ConnectOption customizes ConnectOptions built from RealtimeConfig.
type ConnectOption func(*connectOptionsBuilder) error

type connectOptionsBuilder struct {
	options *ConnectOptions
	query   url.Values
}

// WithQueryParam appends one query parameter to the realtime transport URL.
func WithQueryParam(key, value string) ConnectOption {
	return func(b *connectOptionsBuilder) error {
		if key == "" {
			return fmt.Errorf("golem-go-client: query parameter key cannot be empty")
		}
		if b.query == nil {
			b.query = url.Values{}
		}
		b.query.Add(key, value)
		return nil
	}
}

// WithQuery appends query parameters to the realtime transport URL.
func WithQuery(values url.Values) ConnectOption {
	return func(b *connectOptionsBuilder) error {
		if b.query == nil {
			b.query = url.Values{}
		}
		for key, list := range values {
			if key == "" {
				return fmt.Errorf("golem-go-client: query parameter key cannot be empty")
			}
			for _, value := range list {
				b.query.Add(key, value)
			}
		}
		return nil
	}
}

// WithTLSClientConfig sets an explicit TLS config for the built ConnectOptions.
func WithTLSClientConfig(cfg *tls.Config) ConnectOption {
	return func(b *connectOptionsBuilder) error {
		b.options.TLSClientConfig = cfg
		return nil
	}
}

// ConnectOptionsFromRealtimeConfig converts realtime config JSON into ConnectOptions.
func ConnectOptionsFromRealtimeConfig(cfg RealtimeConfig, opts ...ConnectOption) (ConnectOptions, error) {
	switch cfg.Transport {
	case TransportWebSocket, TransportWebTransport:
	default:
		return ConnectOptions{}, fmt.Errorf("golem-go-client: unsupported transport %q", cfg.Transport)
	}
	options := ConnectOptions{
		Transport:               cfg.Transport,
		URL:                     cfg.URL,
		ServerCertificateHashes: append([]CertificateHash(nil), cfg.ServerCertificateHashes...),
		EventualAckIntervalMs:   cfg.EventualAckIntervalMs,
	}
	builder := &connectOptionsBuilder{options: &options}
	for _, opt := range opts {
		if opt == nil {
			continue
		}
		if err := opt(builder); err != nil {
			return ConnectOptions{}, err
		}
	}
	if len(builder.query) > 0 {
		u, err := url.Parse(options.URL)
		if err != nil {
			return ConnectOptions{}, fmt.Errorf("golem-go-client: parsing realtime transport URL: %w", err)
		}
		query := u.Query()
		for key, list := range builder.query {
			for _, value := range list {
				query.Add(key, value)
			}
		}
		u.RawQuery = query.Encode()
		options.URL = u.String()
	}
	return options, nil
}

// TLSClientConfigFromCertificateHashes builds a TLS config pinned to certificate hashes.
func TLSClientConfigFromCertificateHashes(serverName string, hashes []CertificateHash) (*tls.Config, error) {
	if len(hashes) == 0 {
		return nil, nil
	}
	if err := validateCertificateHashes(hashes); err != nil {
		return nil, err
	}
	pinned := append([]CertificateHash(nil), hashes...)
	return &tls.Config{
		ServerName:         serverName,
		InsecureSkipVerify: true,
		VerifyPeerCertificate: func(rawCerts [][]byte, _ [][]*x509.Certificate) error {
			return verifyPinnedPeerCertificate(serverName, pinned, rawCerts)
		},
	}, nil
}

func decodeCertificateHash(algorithm, value string) (CertificateHash, error) {
	algorithm = normalizeCertificateHashAlgorithm(algorithm)
	if algorithm != certificateHashSHA256 {
		return CertificateHash{}, fmt.Errorf("golem-go-client: unsupported certificate hash algorithm %q", algorithm)
	}
	decoded, err := hex.DecodeString(value)
	if err != nil {
		return CertificateHash{}, fmt.Errorf("golem-go-client: decoding certificate hash: %w", err)
	}
	hash := CertificateHash{Algorithm: algorithm, Value: decoded}
	if err := validateCertificateHashes([]CertificateHash{hash}); err != nil {
		return CertificateHash{}, err
	}
	return hash, nil
}

func validateCertificateHashes(hashes []CertificateHash) error {
	for _, hash := range hashes {
		algorithm := normalizeCertificateHashAlgorithm(hash.Algorithm)
		switch algorithm {
		case certificateHashSHA256:
			if len(hash.Value) != sha256.Size {
				return fmt.Errorf("golem-go-client: %s certificate hash length %d, want %d", certificateHashSHA256, len(hash.Value), sha256.Size)
			}
		default:
			return fmt.Errorf("golem-go-client: unsupported certificate hash algorithm %q", hash.Algorithm)
		}
	}
	return nil
}

func verifyPinnedPeerCertificate(serverName string, hashes []CertificateHash, rawCerts [][]byte) error {
	if len(rawCerts) == 0 {
		return fmt.Errorf("golem-go-client: server sent no certificates")
	}
	leaf, err := x509.ParseCertificate(rawCerts[0])
	if err != nil {
		return fmt.Errorf("golem-go-client: parsing server certificate: %w", err)
	}
	if !matchesCertificateHash(rawCerts[0], hashes) {
		return fmt.Errorf("golem-go-client: server certificate did not match configured hashes")
	}

	intermediates := x509.NewCertPool()
	for _, raw := range rawCerts[1:] {
		cert, err := x509.ParseCertificate(raw)
		if err != nil {
			return fmt.Errorf("golem-go-client: parsing server certificate chain: %w", err)
		}
		intermediates.AddCert(cert)
	}
	verifyOptions := x509.VerifyOptions{
		DNSName:       serverName,
		Intermediates: intermediates,
		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
	}
	if _, err := leaf.Verify(verifyOptions); err == nil {
		return nil
	}

	if serverName != "" {
		if err := leaf.VerifyHostname(serverName); err != nil {
			return fmt.Errorf("golem-go-client: pinned server certificate hostname mismatch: %w", err)
		}
	}
	now := time.Now()
	if now.Before(leaf.NotBefore) || now.After(leaf.NotAfter) {
		return fmt.Errorf("golem-go-client: pinned server certificate is not valid at %s", now.Format(time.RFC3339))
	}
	if !certificateAllowsServerAuth(leaf) {
		return fmt.Errorf("golem-go-client: pinned server certificate is not valid for server auth")
	}
	return nil
}

func matchesCertificateHash(der []byte, hashes []CertificateHash) bool {
	sum := sha256.Sum256(der)
	for _, hash := range hashes {
		if normalizeCertificateHashAlgorithm(hash.Algorithm) != certificateHashSHA256 {
			continue
		}
		if subtle.ConstantTimeCompare(sum[:], hash.Value) == 1 {
			return true
		}
	}
	return false
}

func certificateAllowsServerAuth(cert *x509.Certificate) bool {
	if len(cert.ExtKeyUsage) == 0 {
		return true
	}
	for _, usage := range cert.ExtKeyUsage {
		if usage == x509.ExtKeyUsageServerAuth || usage == x509.ExtKeyUsageAny {
			return true
		}
	}
	return false
}

func normalizeCertificateHashAlgorithm(algorithm string) string {
	return strings.ToLower(strings.TrimSpace(algorithm))
}

func serverNameFromEndpointURL(endpoint string) (string, error) {
	u, err := url.Parse(endpoint)
	if err != nil {
		return "", fmt.Errorf("golem-go-client: parsing WebTransport URL: %w", err)
	}
	if u.Hostname() == "" {
		return "", fmt.Errorf("golem-go-client: WebTransport URL %q has no host", endpoint)
	}
	return u.Hostname(), nil
}

File: golem-js/src/bootstrap.ts
import type {
  ConnectOptions,
  ConnectOptionsResolver,
  ConnectPlan,
  RealtimeEndpoint,
  TransportKind,
  WebTransportCertificateHash,
} from "./client.js";

const certificateHashSHA256 = "sha-256";
const maxErrorBodyBytes = 512;
const maxConfigBodyBytes = 64 * 1024;
const maxInt32 = 0x7fffffff;

function redactURLForError(value: string): string {
  try {
    const parsed = new URL(value);
    return `${parsed.protocol}//${parsed.host}${parsed.pathname}`;
  } catch {
    return "<invalid-url>";
  }
}

function sanitizeFetchErrorText(value: unknown, endpoint: string): string {
  let text = value instanceof Error ? value.message : String(value);
  text = text
    .replace(
      /\b(?:https?|wss?):\/\/[^\s"'<>]+/gi,
      (url) => redactURLForError(url),
    )
    .replace(
      /([?&][A-Za-z0-9_.~-]+)=([^&#\s"'<>]*)/g,
      "$1=<redacted>",
    )
    .replace(
      /\b(ticket|token|access_token|authorization|auth|session|secret|api_key)\s*([=:])\s*[^\s,;"']+/gi,
      "$1$2<redacted>",
    )
    .replace(
      /(["']?(?:ticket|token|access_token|authorization|auth|session|secret|api_key)["']?\s*:\s*["']?)[^"',}\s]+/gi,
      "$1<redacted>",
    );

  try {
    const parsed = new URL(endpoint);
    const values = [parsed.username, parsed.password];
    for (const value of parsed.searchParams.values()) {
      values.push(value, encodeURIComponent(value));
    }
    for (const raw of parsed.search.slice(1).split("&")) {
      const equals = raw.indexOf("=");
      if (equals >= 0) {
        values.push(raw.slice(equals + 1));
      }
    }
    for (const secret of values) {
      if (secret !== "") {
        text = text.split(secret).join("<redacted>");
      }
    }
  } catch {
    // The endpoint itself is never included when it cannot be parsed.
  }
  return text;
}

/** Wire shape served by golem.Server.RealtimeConfigHandler (hex certificate hash values). */
export interface RealtimeConfig {
  transport: TransportKind;
  url: string;
  serverCertificateHashes?: WebTransportCertificateHash[];
  eventualAckIntervalMs?: number;
  fallback?: RealtimeEndpoint;
}

/**
 * Injectable fetch knobs for fetchRealtimeConfig (defaults to global fetch).
 * `signal` takes precedence over `init.signal` when both are set.
 */
export interface FetchRealtimeConfigOptions {
  fetch?: typeof fetch;
  init?: RequestInit;
  signal?: AbortSignal;
}

type ConnectOptionsMutator = (options: ConnectOptions, query: URLSearchParams) => void;

/**
 * Customizes ConnectOptions built from RealtimeConfig.
 * Prefer withQueryParam / withQuery for auth query params (e.g. token, char_id).
 */
export type ConnectOption = ConnectOptionsMutator;

interface realtimeConfigResponse {
  transport?: unknown;
  url?: unknown;
  serverCertificateHashes?: unknown;
  eventualAckIntervalMs?: unknown;
  fallback?: unknown;
}

/**
 * Loads and decodes a realtime config JSON endpoint.
 * Success bodies are capped at 64 KiB; error snippets at 512 bytes via streaming reads.
 * Never includes caller-supplied query token values in thrown errors.
 *
 * Phaser may return connectPlanFromRealtimeConfig's result from connectionOptions;
 * credentials should remain in the plan's lazy per-candidate resolver.
 */
export async function fetchRealtimeConfig(
  endpoint: string,
  options?: FetchRealtimeConfigOptions,
): Promise<RealtimeConfig> {
  const fetchImpl = options?.fetch ?? globalThis.fetch;
  if (typeof fetchImpl !== "function") {
    throw new Error("golem-js: fetch is not available; pass options.fetch");
  }
  const init = mergeFetchInit(options);
  let response: Response;
  try {
    response = await fetchImpl(endpoint, init);
  } catch (err) {
    const message = sanitizeFetchErrorText(err, endpoint);
    throw new Error(`golem-js: fetching realtime config: ${message}`);
  }
  if (!response.ok) {
    const snippet = await readBoundedText(response, maxErrorBodyBytes);
    if (snippet.length > 0) {
      const safeSnippet = sanitizeFetchErrorText(snippet, endpoint);
      throw new Error(`golem-js: fetching realtime config: status ${response.status} body=${JSON.stringify(safeSnippet)}`);
    }
    throw new Error(`golem-js: fetching realtime config: status ${response.status}`);
  }
  const { text, truncated } = await readBoundedTextWithFlag(response, maxConfigBodyBytes);
  if (truncated) {
    throw new Error(`golem-js: realtime config response exceeds ${maxConfigBodyBytes} bytes`);
  }
  let body: realtimeConfigResponse;
  try {
    body = JSON.parse(text) as realtimeConfigResponse;
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    throw new Error(`golem-js: decoding realtime config: ${message}`);
  }
  return parseRealtimeConfig(body);
}

/** Appends one query parameter to the realtime transport URL. */
export function withQueryParam(key: string, value: string): ConnectOption {
  return (_options, query) => {
    if (key === "") {
      throw new Error("golem-js: query parameter key cannot be empty");
    }
    if (typeof value !== "string") {
      throw new Error("golem-js: query parameter values must be strings");
    }
    query.append(key, value);
  };
}

/** Appends query parameters to the realtime transport URL. Duplicate keys are preserved in order. */
export function withQuery(values: Record<string, string | string[]> | URLSearchParams): ConnectOption {
  return (_options, query) => {
    if (values instanceof URLSearchParams) {
      for (const [key, value] of values.entries()) {
        if (key === "") {
          throw new Error("golem-js: query parameter key cannot be empty");
        }
        query.append(key, value);
      }
      return;
    }
    for (const [key, raw] of Object.entries(values)) {
      if (key === "") {
        throw new Error("golem-js: query parameter key cannot be empty");
      }
      if (Array.isArray(raw)) {
        for (const value of raw) {
          if (typeof value !== "string") {
            throw new Error("golem-js: query parameter values must be strings");
          }
          query.append(key, value);
        }
      } else if (typeof raw !== "string") {
        throw new Error("golem-js: query parameter values must be strings");
      } else {
        query.append(key, raw);
      }
    }
  };
}

/**
 * Converts realtime config JSON into a fresh ConnectOptions value.
 * Preserves existing URL query parameters, appends supplied query values with URL encoding,
 * clones certificate hash arrays, and propagates eventualAckIntervalMs.
 *
 * Phaser: call after fetchRealtimeConfig, cache the returned options, and return them from
 * connectionOptions() so reconnects stay synchronous.
 */
export function connectOptionsFromRealtimeConfig(cfg: RealtimeConfig, ...opts: ConnectOption[]): ConnectOptions {
  switch (cfg.transport) {
    case "websocket":
    case "webtransport":
      break;
    default:
      throw new Error(`golem-js: unsupported transport ${JSON.stringify(cfg.transport)}`);
  }
  if (typeof cfg.url !== "string" || cfg.url.trim() === "") {
    throw new Error("golem-js: realtime config url is required");
  }
  if (cfg.eventualAckIntervalMs !== undefined) {
    assertEventualAckIntervalMs(cfg.eventualAckIntervalMs);
  }

  const options: ConnectOptions = {
    transport: cfg.transport,
    url: cfg.url,
    serverCertificateHashes: cfg.serverCertificateHashes?.map(cloneCertificateHash),
    eventualAckIntervalMs: cfg.eventualAckIntervalMs,
  };

  const query = new URLSearchParams();
  for (const opt of opts) {
    if (opt == null) {
      continue;
    }
    opt(options, query);
  }
  if (query.toString() !== "") {
    let parsed: URL;
    try {
      parsed = new URL(options.url);
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      throw new Error(`golem-js: parsing realtime transport URL: ${message}`);
    }
    for (const [key, value] of query.entries()) {
      parsed.searchParams.append(key, value);
    }
    options.url = parsed.toString();
  }
  return options;
}

/**
 * Converts realtime config JSON into an ordered, credential-free connection plan.
 * The optional resolver is invoked lazily before each physical transport dial.
 */
export function connectPlanFromRealtimeConfig(
  cfg: RealtimeConfig,
  resolveOptions?: ConnectOptionsResolver,
): ConnectPlan {
  const primary = connectOptionsFromRealtimeConfig(cfg);
  const candidates: RealtimeEndpoint[] = [primary];
  if (cfg.fallback != null) {
    if (
      primary.transport !== "webtransport" ||
      cfg.fallback.transport !== "websocket"
    ) {
      throw new Error(
        "golem-js: realtime config fallback must be webtransport then websocket",
      );
    }
    const fallback = connectOptionsFromRealtimeConfig({
      transport: cfg.fallback.transport,
      url: cfg.fallback.url,
    });
    candidates.push(fallback);
  }
  return { candidates, resolveOptions };
}

/** Reads a Response body stream up to maxBytes; cancels the reader when the cap is hit. */
export async function readBoundedText(response: Response, maxBytes: number): Promise<string> {
  const { text } = await readBoundedTextWithFlag(response, maxBytes);
  return text;
}

function mergeFetchInit(options?: FetchRealtimeConfigOptions): RequestInit | undefined {
  if (options?.init == null && options?.signal == null) {
    return undefined;
  }
  const init: RequestInit = options?.init ? { ...options.init } : {};
  if (options?.signal != null) {
    init.signal = options.signal;
  }
  return init;
}

function parseRealtimeConfig(body: realtimeConfigResponse): RealtimeConfig {
  if (typeof body.transport !== "string" || body.transport.trim() === "") {
    throw new Error("golem-js: realtime config transport is required");
  }
  if (typeof body.url !== "string" || body.url.trim() === "") {
    throw new Error("golem-js: realtime config url is required");
  }
  const transport = body.transport.trim() as TransportKind;
  switch (transport) {
    case "websocket":
    case "webtransport":
      break;
    default:
      throw new Error(`golem-js: unsupported transport ${JSON.stringify(body.transport)}`);
  }

  const cfg: RealtimeConfig = {
    transport,
    url: body.url,
  };
  if (body.eventualAckIntervalMs !== undefined && body.eventualAckIntervalMs !== null) {
    cfg.eventualAckIntervalMs = assertEventualAckIntervalMs(body.eventualAckIntervalMs);
  }
  if (body.serverCertificateHashes != null) {
    if (!Array.isArray(body.serverCertificateHashes)) {
      throw new Error("golem-js: serverCertificateHashes must be an array");
    }
    cfg.serverCertificateHashes = body.serverCertificateHashes.map((entry, index) => {
      if (entry == null || typeof entry !== "object") {
        throw new Error(`golem-js: serverCertificateHashes[${index}] must be an object`);
      }
      const hash = entry as { algorithm?: unknown; value?: unknown };
      if (typeof hash.algorithm !== "string" || typeof hash.value !== "string") {
        throw new Error(`golem-js: serverCertificateHashes[${index}] requires string algorithm and value`);
      }
      return decodeCertificateHash(hash.algorithm, hash.value);
    });
  }
  if (body.fallback != null) {
    if (typeof body.fallback !== "object" || Array.isArray(body.fallback)) {
      throw new Error("golem-js: realtime config fallback must be an object");
    }
    const fallback = body.fallback as {
      transport?: unknown;
      url?: unknown;
      serverCertificateHashes?: unknown;
      eventualAckIntervalMs?: unknown;
    };
    if (typeof fallback.transport !== "string" || fallback.transport.trim() === "") {
      throw new Error("golem-js: realtime config fallback transport is required");
    }
    if (typeof fallback.url !== "string" || fallback.url.trim() === "") {
      throw new Error("golem-js: realtime config fallback url is required");
    }
    const fallbackTransport = fallback.transport.trim() as TransportKind;
    if (transport !== "webtransport" || fallbackTransport !== "websocket") {
      throw new Error(
        "golem-js: realtime config fallback must be webtransport then websocket",
      );
    }
    if (
      fallback.serverCertificateHashes !== undefined ||
      fallback.eventualAckIntervalMs !== undefined
    ) {
      throw new Error(
        "golem-js: websocket fallback cannot include WebTransport settings",
      );
    }
    cfg.fallback = {
      transport: fallbackTransport,
      url: fallback.url,
    };
  }
  return cfg;
}

function assertEventualAckIntervalMs(value: unknown): number {
  if (typeof value !== "number" || !Number.isFinite(value)) {
    throw new Error("golem-js: eventualAckIntervalMs must be a finite integer");
  }
  if (!Number.isInteger(value)) {
    throw new Error("golem-js: eventualAckIntervalMs must be a finite integer");
  }
  if (value < 0 || value > maxInt32) {
    throw new Error("golem-js: eventualAckIntervalMs out of int32 range");
  }
  return value;
}

function decodeCertificateHash(algorithm: string, value: string): WebTransportCertificateHash {
  const normalizedAlgorithm = algorithm.trim().toLowerCase();
  if (normalizedAlgorithm !== certificateHashSHA256) {
    throw new Error(`golem-js: unsupported certificate hash algorithm ${JSON.stringify(algorithm)}`);
  }
  const bytes = hexToBytes(value);
  if (bytes.byteLength !== 32) {
    throw new Error(`golem-js: sha-256 certificate hash length ${bytes.byteLength}, want 32`);
  }
  return { algorithm: normalizedAlgorithm, value };
}

function hexToBytes(hex: string): Uint8Array {
  const normalized = hex.replace(/\s+/g, "");
  if (normalized.length % 2 !== 0) {
    throw new Error("golem-js: certificate hash hex string must have even length");
  }
  if (!/^[0-9a-fA-F]*$/.test(normalized)) {
    throw new Error("golem-js: certificate hash hex string contains non-hex characters");
  }
  const out = new Uint8Array(normalized.length / 2);
  for (let i = 0; i < normalized.length; i += 2) {
    out[i / 2] = Number.parseInt(normalized.slice(i, i + 2), 16);
  }
  return out;
}

function cloneCertificateHash(hash: WebTransportCertificateHash): WebTransportCertificateHash {
  if (typeof hash.value === "string") {
    return { algorithm: hash.algorithm, value: hash.value };
  }
  if (ArrayBuffer.isView(hash.value)) {
    const view = hash.value;
    const copy = new Uint8Array(view.byteLength);
    copy.set(new Uint8Array(view.buffer, view.byteOffset, view.byteLength));
    return { algorithm: hash.algorithm, value: copy.buffer };
  }
  const copy = new Uint8Array(hash.value.byteLength);
  copy.set(new Uint8Array(hash.value));
  return { algorithm: hash.algorithm, value: copy.buffer };
}

async function readBoundedTextWithFlag(
  response: Response,
  maxBytes: number,
): Promise<{ text: string; truncated: boolean }> {
  const body = response.body;
  if (body && typeof body.getReader === "function") {
    const reader = body.getReader();
    const chunks: Uint8Array[] = [];
    let total = 0;
    let truncated = false;
    try {
      while (true) {
        const { done, value } = await reader.read();
        if (done) {
          break;
        }
        if (!value || value.byteLength === 0) {
          continue;
        }
        if (total + value.byteLength > maxBytes) {
          const take = maxBytes - total;
          if (take > 0) {
            chunks.push(value.subarray(0, take));
            total = maxBytes;
          }
          truncated = true;
          try {
            await reader.cancel();
          } catch {
            // ignore cancel failures
          }
          break;
        }
        chunks.push(value);
        total += value.byteLength;
      }
    } catch {
      return fallbackBoundedText(response, maxBytes);
    } finally {
      try {
        reader.releaseLock();
      } catch {
        // already released/canceled
      }
    }
    return { text: decodeChunks(chunks, total), truncated };
  }
  return fallbackBoundedText(response, maxBytes);
}

async function fallbackBoundedText(
  response: Response,
  maxBytes: number,
): Promise<{ text: string; truncated: boolean }> {
  try {
    if (typeof response.arrayBuffer === "function") {
      const buffer = await response.arrayBuffer();
      const truncated = buffer.byteLength > maxBytes;
      const bytes = new Uint8Array(buffer, 0, Math.min(buffer.byteLength, maxBytes));
      return { text: new TextDecoder().decode(bytes), truncated };
    }
    const text = await response.text();
    if (text.length > maxBytes) {
      return { text: text.slice(0, maxBytes), truncated: true };
    }
    return { text, truncated: false };
  } catch {
    return { text: "", truncated: false };
  }
}

function decodeChunks(chunks: Uint8Array[], total: number): string {
  const merged = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return new TextDecoder().decode(merged);
}
